text/0000-attribute-actions.md

```diff
@@ -0,0 +1,244 @@
+
+Passing non-actions to `on*` attributes will be discouraged.
+
+* Literals should be permitted, such as `onclick="window.alert('foo')"`.
+* Non-function bindings are not permitted, and will throw from an assertion.
+  For example `onclick={{someBoundString}}` will cause an assertion to throw.
+* Non-action functions such as `onclick={{someBoundFunction}}` should cause
+  a warning to be logged. This warning will encourage the user to use
+  `onclick={{action someBoundFunction}}` or use actions as normally
+  documented.
+
+#### Handling an Attribute Action
+
+`on*` attributes will be invoked with the raw DOM event (not the jQuery event)
+as an argument. For example:
+
+```js
+// app/components/my-button.js
+export Ember.Component.extend({
+  actions: {
+    save(event) {
+      console.log('saved with click on ', event.target);
+      this.get('model').save();
+    }
+  }
+});
+```
+
+```hbs
+{{! app/templates/components/my-button.hbs }}
+<button onclick={{action 'save'}}>Save</button>
+```
+
+Actions create a closure over arguments, thus the event may not always be
+the first argument. For example:
+
+```js
+// app/components/my-input.js
+export Ember.Component.extend({
+  actions: {
+    log(prefix, event) {
+      console.log(prefix, Ember.$(event.target).val());
+    }
+  }
+});
+```
+
+```hbs
+{{! app/templates/components/my-input.hbs }}
+<input onchange={{action 'log' 'value is:'}} />
+```
+
+Additionally, the `value` argument to `action` is useful when passing the
+event.
+
+
+```js
+// app/components/my-input.js
+export Ember.Component.extend({
+  actions: {
+    log(value) {
+      console.log(target);
+    }
+  }
+});
+```
+
+```hbs
+{{! app/templates/components/my-input.hbs }}
+<input onchange={{action 'log' value='target.value'}} />
+```
+
+Passing the event is useful for interacting with the originating DOM node,
+but additionally important for allowing event propagation to be controlled
+(cancelled).
+
+#### Event Dispatching
+
+Attribute actions utilize a native API for dispatching that places them on
+the bubbling event dispatcher. Browsers supported by Ember 2.x have three
+kind of DOM event attachment:
+
+* `el.addEventListener('click', handlerFn, false)` add a handler to the bubbling
+  phase.
+* `el.addEventListener('click', handlerFn, true)` add a handler to the capture
+  phase.
+* `el.onclick = handlerFn;` add a handler to the bubbling phase.
+
+Attribute actions will be dispatched on the bubbling phase, and attached via
+`addEventListener`. This ensures that if there are multiple handlers making
+their way onto an element they do not stomp on each other. Any attribute named
+`on*` will be attached as such.
+
+For a webcomponent to dispatch an attached attribute action, it should use
+`dispatchEvent`.
+
+Glimmer components may (and this is specilation more than design) permit
+multiple events to be attached via reflection of invocation attrs to the
+root element. For example:
```

Comment:
I doubt we can get consensus on anything regarding glimmer components that is definitive. However I also remain strongly in favor of introducing an improved API for actions. The current situation is terrible:

* Idiomatic usage includes both bubbling and non-bubbling forms of the action helper.
* Many early adopters have started using `onclick={{action`. This API is not documented or designed at all.

To continue the status quo until glimmer components land does not seem wise.

I agree we should take various designs for glimmer component reflection and merging into account, but I'm not comfortable writing those into stone for this RFC. I'm trying to strike a balance here by specifying this is only speculation.